### PR TITLE
Improve the Byzantine validator detector

### PR DESCRIPTION
**Pull request by nbougalis**
This commit expands the detection capabilities of the Byzantine validation detector originally introduced with commit 381606aba2efcd179517f2497dc6e8c2770c679e. Prior to this change only validators that were on a server's UNL were monitored. Now, all the validations that a server receives are passed through the detector.

### CODE REVIEW

**Comment by seelabs** [comment]
@nbougalis A couple of the validation tests and manifest tests are failing:
```
21> #14 failed: Manifest_test.cpp(990)
21> #15 failed: Manifest_test.cpp(993)
21> #15 failed: Manifest_test.cpp(481)
8> #6 failed: Validations_test.cpp(306)
8> #7 failed: Validations_test.cpp(309)
```

I looked at the first one. It's now returning `badEphemeralKey`, which is correct. Looks like we just need to update the test. I didn't look at the others.

File: src/ripple/app/misc/NetworkOPs.cpp
Comment (gregtatcam) [inline src/ripple/app/misc/NetworkOPs.cpp:2510]: I think this is useful for debugging, especially with validation/proposal reduce-relay feature enabled. Maybe just change it to `trace` to reduce the noise?

File: src/ripple/app/consensus/RCLConsensus.h
Comment (seelabs) [inline src/ripple/app/consensus/RCLConsensus.h:67]: The advantage of the old code is we don't have to worry about lifetime issues. I code is correct - the value that is passed in lives in the application object, which should outlive this. Still, it makes the code a little more fragile.

I probably weakly prefer the old code, but fine as-is as well.

File: src/ripple/overlay/impl/PeerImp.cpp
Comment (seelabs) [inline src/ripple/overlay/impl/PeerImp.cpp:2252]: Is there a reason this was moved inside the try block? I don't believe `closeTime` or the other functions can throw.

I don't object, but in this same file we move some code outside of a `try` block. I'm not sure the reason this was was moved in while the other one was moved out.

File: src/ripple/overlay/impl/PeerImp.cpp
Comment (seelabs) [inline src/ripple/overlay/impl/PeerImp.cpp:2691]: See comment above: I'm not sure why this was moved outside the `try` block while the other was moved in.

**Review by seelabs — COMMENTED** [review]
Looks good. Left some minor comments. Will wait until the unit tests pass before I give a thumbs up.

File: src/ripple/app/consensus/RCLValidations.cpp
Comment (bachase) [inline src/ripple/app/consensus/RCLValidations.cpp:174]: It's been a while since I've looked at this, can you clarify this comment being unchanged?

> masterKey is seated only if validator is trusted or listed

Your comment https://github.com/ripple/rippled/pull/3778#issue-578184333 suggested we only previously checked trusted validations.

The

**Review by gregtatcam — COMMENTED** [review]
It looks good to me. I'll wait for the unittest fixes to approve.

File: src/ripple/app/misc/NetworkOPs.cpp
Comment (nbougalis) [reply src/ripple/app/misc/NetworkOPs.cpp:2510]: :+1:

File: src/ripple/app/consensus/RCLConsensus.h
Comment (nbougalis) [reply src/ripple/app/consensus/RCLConsensus.h:67]: The old code makes it more difficult to 're-key' a server (e.g. by allowing it to switch its manifest) although that's not something currently supported. It also dumps keying material in whatever random memory location the `this` exists in which makes it more difficult to "secure" the memory for sensitive keying material (e.g. by marking it as "do not include in core dump" and by other similar, high-security techniques, including, but not limited do storing the key itself inside an HSM).

I understand the concern here, but the code is structured so that the object being passed in a singleton. I guess we could _reinforce_ that by passing a `std::shared_ptr` or somesuch.

File: src/ripple/overlay/impl/PeerImp.cpp
Comment (nbougalis) [reply src/ripple/overlay/impl/PeerImp.cpp:2252]: No, it can be moved out.

File: src/ripple/app/consensus/RCLValidations.cpp
Comment (nbougalis) [reply src/ripple/app/consensus/RCLValidations.cpp:174]: This is somewhat confusing, and I understand why you'd call this out. The comment is still accurate (sorta-kinda): we have a validation signed with some key K which _may_ be an ephemeral key or it may be a "direct" key (i.e. set via `[validation_seed]`). If it's signed by a manifest **AND** we know the manifest, we want to retrieve the master key and use that, otherwise, we use the ephemeral key.

Operating with "direct" keys (i.e. without a manifest) should be deprecated and, eventually, not allowed. And, if we receive a validation for an ephemeral key whose manifest the server does not have, we ought to just ignore it, especially post 1.7 that fixed manifest propagation.

**Review by bachase — APPROVED** [review]
LGTM (assuming other comments on unit tests are addressed).

**Review by seelabs — APPROVED** [review]
:+1:

**Pull request closed without merging**
